apiml-security-common/src/main/java/org/zowe/apiml/security/common/verify/TrustedCertificatesProvider.java

```diff
@@ -0,0 +1,114 @@
+                                       @Qualifier("publicKeyCertificatesBase64") Set<String> publicKeyCertificatesBase64) {
+        this.httpClient = httpClient;
+        this.publicKeyCertificatesBase64 = publicKeyCertificatesBase64;
+    }
+
+    /**
+     * Query given rest endpoint to get the certificate chain from remote proxy gateway.
+     * The endpoint should be publicly available and should provide the certificate chain in PEM format.
+     *
+     * @param certificatesEndpoint Given full URL to the remote proxy gateway certificates endpoint
+     * @return List of certificates or empty list
+     */
+    @Cacheable(value = "trustedCertificates", key = "#certificatesEndpoint", unless = "#result.isEmpty()")
+    public List<Certificate> getTrustedCerts(String certificatesEndpoint) {
+        List<Certificate> trustedCerts = new ArrayList<>();
+        String pem = callCertificatesEndpoint(certificatesEndpoint);
+        if (StringUtils.isNotEmpty(pem)) {
+            try {
+                Collection<? extends Certificate> certs = CertificateFactory
+                    .getInstance("X.509")
+                    .generateCertificates(new ByteArrayInputStream(pem.getBytes()));
+                trustedCerts.addAll(certs);
+                updateTrustedPublicKeys(trustedCerts);
+            } catch (CertificateException e) {
+                apimlLog.log("org.zowe.apiml.security.common.verify.errorParsingCertificates", e.getMessage());
+            }
+        }
+        return trustedCerts;
+    }
+
+    private String callCertificatesEndpoint(String url) {
+        try {
+            HttpGet httpGet = new HttpGet(new URI(url));
+            HttpResponse httpResponse = httpClient.execute(httpGet);
+            final int statusCode = httpResponse.getStatusLine() != null ? httpResponse.getStatusLine().getStatusCode() : 0;
+            String body = "";
+            if (httpResponse.getEntity() != null) {
+                body = EntityUtils.toString(httpResponse.getEntity(), StandardCharsets.UTF_8);
+            }
+            if (statusCode != HttpStatus.SC_OK) {
+                apimlLog.log("org.zowe.apiml.security.common.verify.invalidResponse", url, statusCode, body);
+                return null;
+            }
+            log.debug("Trusted certificates from {}: {}", url, body);
+            return body;
+
+        } catch (URISyntaxException e) {
+            apimlLog.log("org.zowe.apiml.security.common.verify.invalidURL", e.getMessage());
+        } catch (IOException e) {
+            apimlLog.log("org.zowe.apiml.security.common.verify.httpError", e.getMessage());
+        }
+        return null;
+    }
+
+    /**
+     * Updates the list of public keys of APIML trusted certificates
+     *
+     * @param certs List of trusted certificates
+     */
+    private void updateTrustedPublicKeys(List<Certificate> certs) {
```

Comment:
this method is rather confusing. It is updating list of certificates that belongs to API ML not all trusted.

Reply:
Refactored and renamed.